rfcs/2021-06-15-shipping-blade.md

```diff
@@ -0,0 +1,438 @@
+export * from 'src/tokens/utils/useColorScheme'
+```
+
+#### Cons
+* We have to write the whole named `import` statement for everything we want to import and with every category
+
+### Which approach we'll take?
+We'll go with approach [#3](#3-import-from-each-categorycomponents-tokenstypes-utils-root-package)
+
+## What strategy to take for packaging?
+
+### 1. Multi-package
+We can have each component as individual packages. Each of them will have their own `package.json` and versioning.
+
+```js
+// package.json.name: "@razorpay-blade/button" 
+import Button from '@razorpay-blade/button'
+
+// package.json.name: "@razorpay-blade/text" 
+import Text from '@razorpay-blade/text'
+
+// package.json.name: "@razorpay-blade/tokens" 
+import { overrideTheme, paymentTheme } from '@razorpay-blade/tokens'
+```
+
+### 2. Mono-package
+We can have mono package i.e only one `package.json` for whole blade
+```js
+// package.json.name: "@razorpay/blade"
+import { Button, Text } from '@razorpay/blade/components'
+import { overrideTheme, paymentTheme } from '@razorpay/blade/tokens'
+```
+
+Let's understand the difference by taking a use case(note: there's no performance concern in either of the styles of packaging)
+
+Payment and Banking teams both start using blade at 1.0.0(assuming we publish a mono package). 
+
+Now, we add more features to blade and bump the version to 1.0.2
+
+➡️ **Banking team** was using 1.0.0 and updated to 1.0.2.
+
+➡️ **Payment team** is still using 1.0.0.
+
+we made fixes/features and couple of releases in between and now our version stands at 1.0.7.
+
+➡️ **Banking team** regularly updated versions and they are at 1.0.7.
+
+➡️ **Payment team** is still using 1.0.0 and now they want to upgrade.
+
+With the mono-package approach once they update to **1.0.7** they need to test their code to ensure nothing is breaking for them(even though we'll ensure we don't do breaking changes but still you never know).
+
+Now, if we go with multi-package version, each component is versioned so **Payment team** can only update the components that they really need to upgrade without worrying about all the components. A lot of big companies follow this approach. Sounds like a good deal right?
+
+Using individually versioned packages does have some downsides that we should look into as it might lead to bloat.
+
+1. Manual upgrades
+
+    * You must manage versioning yourself. Unlike the mono-package, multi-package versioned packages must be manually upgraded to ensure compatibility between components. Upgrading a component that depends on another component may require that component to also be updated to ensure compatibility.
+    * Assume Payment's team is using `Modal` component and they need to just upgrade `Modal`.
+    * But, `Modal` might have other components like `Button`, `Icon` and `Theme` which also would have undergone changes so they need to bump those packages too, now `Button` might depend on `Text` so they need to update `Text` too.
```

Comment:
It depends on how the Modal is adding other packages as a dependency. We can do follow caret ^ to tilde ~ and let npm take care of resolution.
If we do semver correctly for packages, the package manage will figure out if the Modal needs Text that has differently supported version than the one that is available in the consumer app's package.json

This might have other caveats like increasing bundle size if people don't upgrade things correctly.
But let's consider and list this approach?